docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,240 @@
+
+## Signals
+
+### Signal definitions
+
+A LiveCode module may include signal type definitions in its top-level scope [1-2]:
+
+    public signal type <name:id>(<params:formals>)
+
+If the module is imported by another module via the `use` syntax, the signal definitions are made available in the same way as a variable, handler or type definition.
+
+**Notes**:
+
+1. These definitions provide a convenient anchor point for documentation comments about the purpose and usage of a signal.
+
+2. It would be helpful to provide a `com.livecode.signals` module in the standard library that provides definitions of "standard" signals and descriptions of the recommended way to use them.  Examples of standard signals might include `Clicked`, `SelectionChanged`, etc.
+
+### Signal emission
+
+Signals are explicitly emitted by a widget using an `emit` statement:
+
+    emit <name:id>(<params:expressions>)
+
+It is a run-time error to emit a signal with the wrong number or types of emission parameters [1].  The compiler may statically verify that signals are emitted with compatible parameter lists.
+
+The `emit` statement may return before all the signal handlers for the signal have finished running [2].
+
+The signal handlers for a signal get run exactly once for each emission.  All of the handlers for an emission must complete before any of the handlers for a subsequent emission begin [3].
+
+The signal handlers for a signal may be run in any order.
+
+The `emit` statement does not return a value [4].
+
+**Notes**:
+
+1. In principle a widget could set properties on itself before emitting a signal.  However, when adding features to LCS previously it's been found that users prefer relevant information to be passed as parameters to messages rather than out-of-band.  Passing parameters to signal emission also helps to ensure that all of the correct setup has been performed for signal handling - for example, if a parameter is added or removed from the signal, it's easy to verify that all relevant code has been updated.
+
+2. Although, initially, signal emission _could_ be implemented by iterating over registered signal handlers and calling them within the execution of the `emit` statement, this could easily lead to arbitrarily-deep signal emission chains going backwards and forwards between LCS and LCB contexts.  By allowing signal emission to be delayed, this is avoided (the current stack can be allowed to unwind before signal dispatch is performed).  Delaying signal handling also helps to reason about the safety of deleting a widget in response to receiving a signal from it.
+
+3. Emitting a signal twice from the same handler - even with the same parameters - will cause all of the handlers for that signal to be run twice.
+
+4. This is a consequence of asynchronous signal handling.
```

Comment:
The fact that `emit signal` does not return a value is a consequence of asynchronous signal handling.